src/Compilers/Core/Portable/CodeGen/PrivateImplementationDetails.cs

```diff
@@ -140,38 +147,119 @@ internal void Freeze()
+        /// all other sizes, a type is generated of the same size as the data, and that type needs its .pack set to
+        /// the alignment required for the underlying data. While that .pack value isn't required by anything else
+        /// in the compiler (the compiler always aligns RVA fields at 8-byte boundaries, which accomodates any
+        /// element type that's relevant), it is necessary for IL rewriters. Such rewriters also need to ensure
+        /// an appropriate alignment is maintained for the RVA field, and while they could also simplify by choosing
+        /// a worst-case alignment as does the compiler, they may instead use the .pack value as the alignment
+        /// to use for that field, since it's an opaque blob with no other indication as to what kind of data is
+        /// stored and what alignment might be required.
+        /// </param>
+        /// <returns>The field. This may have been newly created or may be an existing field previously created for the same data and alignment.</returns>
+        internal Cci.IFieldReference CreateDataField(ImmutableArray<byte> data, ushort alignment)
         {
             Debug.Assert(!IsFrozen);
-            Cci.ITypeReference type = _proxyTypes.GetOrAdd((uint)data.Length, GetStorageStruct);
-            return _mappedFields.GetOrAdd(data, data0 =>
+
+            Cci.ITypeReference type = _proxyTypes.GetOrAdd(
+                ((uint)data.Length, Alignment: alignment), key =>
+                key.Size switch
+                {
+                    1 when _systemInt8Type is not null => _systemInt8Type,
+                    2 when _systemInt16Type is not null => _systemInt16Type,
+                    4 when _systemInt32Type is not null => _systemInt32Type,
+                    8 when _systemInt64Type is not null => _systemInt64Type,
+                    _ => new ExplicitSizeStruct(key.Size, key.Alignment, this, _systemValueType)
+                });
+
+            return _mappedFields.GetOrAdd((data, alignment), key =>
             {
-                var name = GenerateDataFieldName(data0);
-                var newField = new MappedField(name, this, type, data0);
-                return newField;
+                // For alignment of 1 (which is used in cases other than in fields for ReadOnlySpan<byte>),
+                // just use the hex value of the data hash.  For other alignments, tack on a '2', '4', or '8'
+                // accordingly.  As every byte will yield two chars, the odd number of chars used for 2/4/8
+                // alignments will never produce a name that conflicts with names for an alignment of 1.
+                Debug.Assert(alignment is 1 or 2 or 4 or 8, $"Unexpected alignment: {alignment}");
+                string hex = HashToHex(key.Data);
+                string name = alignment switch
+                {
+                    2 => $"{hex}2",
```

Comment:
>$"{hex}2" [](http://example.com/codeflow?start=25&length=9)

What's the deal with interpolation? Can't we use a regular plus?

Reply:
We could do hex + "2". Why do you prefer it? I generally prefer interpolation because it'll generally always do the best thing now, at least as of .NET 6. In this case, it'll be lowered identically, so if there's a style preference for concatenation, fine.

Reply:
In general, interpolations are harder to read. One should almost become a compiler to parse it. If intent is to concatenate two string, there is nothing better than `+`.

Reply:
> interpolations are harder to read. One should almost become a compiler to parse it.

That's obviously a personal opinion and one I don't share. But I'll change it.